File: dbxml/examples/java/gettingStarted/updateDocument.java
//
// See the file LICENSE for redistribution information.
//
// Copyright (c) 2004,2009 Oracle.  All rights reserved.
//
//

package dbxml.gettingStarted;

import java.io.*;

import com.sleepycat.dbxml.*;

class updateDocument
{

    private static String theContainer = "namespaceExampleData.dbxml";
    private static String collection = "collection('namespaceExampleData.dbxml')";
    private static void usage() {
	String usageMessage = "\nThis program updates a document stored in a DB XML container.\n";
	usageMessage += "You should run exampleLoadContainer before running this example.\n";
	usageMessage += "You are only required to pass this command the path location of the\n";
	usageMessage += "database environment that you specified when you loaded the examples data:\n\n";

	usageMessage += "\t-h <dbenv directory>\n";

	usageMessage += "For example:\n";
	usageMessage += "\tjava com.sleepycat.dbxml.examples.gettingStarted.updateDocument -h examplesEnvironment\n";

	System.out.println(usageMessage);
	System.exit( -1 );
    }

    //Modifies an XML document that is stored in a DB XML container
    private static void doUpdateDocument(XmlManager mgr, XmlContainer container,
					 String query, XmlQueryContext context,
					 XmlTransaction txn )
	throws Throwable {
	String fullQuery = collection + query;
	System.out.println("Updating documents for expression: '" + fullQuery + "'.");
	System.out.println("Return to continue: ");
	System.in.read();

	//query for all the documents that we want to update
	XmlResults results = mgr.query(txn, fullQuery, context, null);
	System.out.println("Found " + results.size() + " matching the expression '"
			   + fullQuery + "'.");

	XmlValue value = results.next();
	while (value != null) {

	    XmlDocument document = value.asDocument();
	    //Retrieve the entire document as a single String object
	    String docString = document.getContentAsString();
	    System.out.println("Updating document: ");
	    System.out.println(docString);

	    //This next method just modifies the document string
	    //in a small way.
	    String newDocString = getNewDocument(mgr, document, context,
						 docString);

	    System.out.println("Updating document...");

	    //Set the document's content to be the new document string
	    document.setContent(newDocString);

	    //Now replace the document in the container
	    container.updateDocument(txn, document);
	    System.out.println("Document updated.");
	    value = results.next();
	}
	results.delete();
    }

    private static String getNewDocument( XmlManager dbXML, XmlDocument document,
					  XmlQueryContext context,
					  String docString )
	throws Throwable {
	//Get the substring that we want to replace
	String inventory = getValue(dbXML, document,
				    "fn:string(/*/inventory/inventory)", context);
	String retVal = new String();

	//Find the substring in the original document string
	int pos = docString.indexOf(inventory);
	if (pos == -1) {
	    System.err.println("Error: inventory string: '" + inventory +
			       "' not found in document string:");
	    System.err.println(docString);

	} else {
	    //Convert the String representation of the inventory level to an
	    // integer, increment by 1, and then convert back to a String for
	    // replacement on the document.
	    int newInventory = Integer.parseInt(inventory) + 1;
	    String newVal = Integer.toString(newInventory);

	    int inventoryLength = inventory.length();
	    int endPos = pos + inventoryLength;

	    StringBuffer strbuff = new StringBuffer(document.getContentAsString());

	    //Perform the replace
	    strbuff.replace(pos, endPos, newVal);
	    System.out.println("Inventory was " + inventory + ", it is now " +
			       newVal + ".");
	    retVal = strbuff.toString();
	}

	return retVal;
    }

    private static String getValue(XmlManager mgr, XmlDocument theDocument,
                                   String query, XmlQueryContext context )
	throws Throwable {

        //Perform the query
	// The document provides the context for the query
        XmlQueryExpression queryExpr = mgr.prepare(query, context);
	XmlValue docVal = new XmlValue(theDocument);
        XmlResults result = queryExpr.execute(docVal, context, null);

        //We require a result set size of 1.
        if (result.size() != 1) {
	    System.err.println("Error! query '" + query +
			       "' returned a result size size != 1");
	    throw new Exception( "getValue found result set not equal to 1.");
        }

        //Get the value. If we allowed the result set to be larger than size 1,
        //we would have to loop through the results, processing each as is
        //required by our application.
        XmlValue value = result.next();
	String retVal = value.asString();
	result.delete();
	queryExpr.delete();

        return retVal;

    }

    //Utility function to clean up objects, exceptions or not,
    // containers and environments must be closed.
    private static void cleanup(myDbEnv env, XmlContainer openedContainer) {
	try {
	    if (openedContainer != null)
		openedContainer.delete();
	    if (env != null)
		env.cleanup();
	} catch (Exception e) {
	    // ignore exceptions on close
	}
    }

    public static void main(String args[])
	throws Throwable {

	File path2DbEnv = null;
	for(int i = 0; i < args.length; ++i) {
	    if (args[i].startsWith("-")) {
		switch(args[i].charAt(1)) {
		case 'h':
		    path2DbEnv = new File( args[++i] );
		    break;
		default:
                    usage();
		}
            }
	}

	if (path2DbEnv == null || ! path2DbEnv.isDirectory()) {
            usage();
	}

	myDbEnv env = null;
	XmlContainer openedContainer = null;
	XmlTransaction txn = null;
	try {
	    // Open an environment
	    env = new myDbEnv(path2DbEnv);
	    XmlManager theMgr = env.getManager();

	    // open a transactional container
	    XmlContainerConfig config = new XmlContainerConfig();
	    config.setTransactional(true);
	    openedContainer =
		theMgr.openContainer(theContainer, config);

	    // Start a transaction
	    txn = theMgr.createTransaction();

            XmlQueryContext context = theMgr.createQueryContext();
            context.setNamespace( "fruits", "http://groceryItem.dbxml/fruits");
            context.setNamespace( "vegetables",
				  "http://groceryItem.dbxml/vegetables");
            context.setNamespace( "desserts", "http://groceryItem.dbxml/desserts");

            //update the document that describes Zapote Blanco (a fruit)
            String query = "/fruits:item[fn:string(product) = 'Zapote Blanco']";
            doUpdateDocument(theMgr, openedContainer, query, context, txn);

            //Commit the writes. This causes the container write operations
            //  to be saved to the container.
            txn.commit();
	} catch (Exception e) {
	    System.err.println("Error performing document update against "
			       + theContainer);
	    System.err.println("   Message: " + e.getMessage());
	    //In the event of an error, we abort the operation
	    // The database is left in the same state as it was in before
	    // we started this operation.
	    if ( txn != null ) {
		txn.abort();
	    }
	    throw e;

	}
	finally {
	    cleanup(env, openedContainer);
	}
    } //End main
}

File: dbxml/examples/perl/gettingStarted/updateDocument.pl


use strict;
use warnings;
use Sleepycat::DbXml 'simple';


sub usage()
{
    warn <<EOM;
This program updates a document found in a DBXML container. You should
pre-load the container using loadExamplesData.[sh|cmd] before running this
example. You are only required to pass this command the path location of the
database environment that you specified when you pre-loaded the examples
data:

    -h <dbenv directory>
EOM

        exit( -1 );
}

sub doUpdateDocument ($$$$$)
{
    my $container = shift ;
    my $query = shift ;
    my $context = shift ;
    my $txn = shift ;
    my $mgr = shift ;
    
    # //////////////////////////////////////////////////////////////////////
    # ////  Updates a document stored in a DBXML container.          ///////
    # //////////////////////////////////////////////////////////////////////
    
    # /// some defensive code eliminated for clarity //
    my $fullQuery = "collection('" . $container->getName() . "')$query";    
    eval {
	print "Updating document for expression: '$query'\n";
	print "Return to continue: ";
	<>;
	
	my $results = $mgr->query($txn, $fullQuery, $context ) ;
	print "Found " . $results->size() 
	    . " documents matching the expression '$fullQuery'\n\n";
	
        # create an update context. This is optional but since we could be performing more than
        #  one update, this could offer us a potential performance benefit.
	my $updateContext = $mgr->createUpdateContext();
	
	my $theDocument = $mgr->createDocument();
	while( $results->next($theDocument) ) {
	    my $docString;
	    
            # Retrieve the document's value as a string.
	    $docString = $theDocument->getContent();
	    
	    print "Updating document: \n$docString\n" ;
	    print "Return to continue: ";
	    <>;
	    
            # Update the document. We do it as a string.
	    
            # This next function just modifies the document string
            # in a small way.
	    $docString = getNewDocument( $mgr, $theDocument, $context, $docString);
	    
	    print "Updating document...\n";
	    
            # Set the document's content to be the new document string
	    $theDocument->setContent( $docString );
	    
            # Now replace the document in the container
	    $container->updateDocument($txn, $theDocument, $updateContext);
	    print "Document updated.\n";
	}
    }; 
    # Catches XmlException
    if (my $e = catch std::exception)
    {
	warn "Document deletion failed. \n";
        warn $e->what() . "\n";
        exit( -1 );
    }
    elsif ($@)
    {
	warn "Document deletion failed. \n";
        warn $@;
        exit( -1 );
    }    
}


sub getValue ($$$$)
{
    my $mgr = shift ;
    my $document = shift ;
    my $query = shift ;
    my $context = shift ;
    
    # ///////////////////////////////////////////////////////////////////////////////
    # /////////    Return specific information from a document. ///////////////////// 
    # /////////   !!!!!! Assumes a result set of size 1 !!!!!!! /////////////////////
    # ///////////////////////////////////////////////////////////////////////////////
    
    
    #  Exception handling omitted....
    
    # Perform the query
    my $doc_expr = $mgr->prepare($query, $context);
    my $result = $doc_expr->execute(new XmlValue($document), $context);

    # We require a result set size of 1.
    die "Expected a result set size of 1"
        unless $result->size == 1 ;

    # Get the value. If we allowed the result set to be larger than size 1,
    # we would have to loop through the results, processing each as is
    # required by our application.
    my $value;
    $result->next($value);

    return $value;
        
}

# Simple little function to replace part of a string with a new value.
# All this does is add an 'A' to the end of the document's inventory
# value. So each time you run this program, the inventory value should
# get longer by one 'A'.
sub getNewDocument( $$$$)
{
    my $mgr = shift ;
    my $theDocument = shift ;
    my $context = shift ;
    my $docString = shift ;

    # get the substring that we want to replace
    my $inventory = getValue($mgr, $theDocument, 
                             '/*/inventory/inventory/text()', $context );

    if ($docString !~ s/($inventory)/${1}A/)
    {
        warn "Error: inventory string: '$inventory' not found in document string:\n";
        warn "$docString\n";
    }

    return $docString;
}

my $path2DbEnv;
my $theContainer = "namespaceExampleData.dbxml";
while (@ARGV && $ARGV[0] =~ /^-/ )
{
    my $opt = shift @ARGV ;

    if ($opt =~ s/^-h//)
      { $path2DbEnv = $opt || shift }
    else
      { usage() }
}

usage()
     if ! $path2DbEnv ;

 # Open a container in the db environment
my $env = new DbEnv(0);
$env->set_cachesize(0, 64 * 1024 * 1024, 1);
$env->open($path2DbEnv,
    Db::DB_INIT_MPOOL|Db::DB_CREATE|Db::DB_INIT_LOCK|Db::DB_INIT_LOG|Db::DB_INIT_TXN);
my $mgr = new XmlManager($env);

my $containerTxn = $mgr->createTransaction();
my $container = $mgr->openContainer($containerTxn, $theContainer);
$containerTxn->commit();

# Get a transaction
my $txn = $mgr->createTransaction();

# Create a context and declare the namespaces
my $context = $mgr->createQueryContext();
$context->setNamespace('fruits'     => 'http://groceryItem.dbxml/fruits');
$context->setNamespace('vegetables' => 'http://groceryItem.dbxml/vegetables');
$context->setNamespace('desserts'   => 'http://groceryItem.dbxml/desserts');

# update the document that describes Zapote Blanco (a fruit)
doUpdateDocument( $container, "/fruits:item/product[text() = 'Zapote Blanco']", 
                      $context, $txn, $mgr );

$txn->commit();

